### PR TITLE
Add expectedCount to Target in listen request

### PR DESCRIPTION
**Pull request by milaGGL**
- [x] add expectedCount to Target
- [x] add expectedCount to spec test builder without exploding current tests.
--spec builders affected: userListens, restoreListen, expectListen, expectActiveTargets
- [x] add simple spec tests to make sure the expectedCount parameter is working as expected

### CODE REVIEW

File: packages/firestore/src/core/sync_engine_impl.ts [hunk 1]
```diff
@@ -316,21 +316,23 @@ export async function syncEngineListen(
...
     targetId = targetData.targetId;
+
     viewSnapshot = await initializeViewAndComputeSnapshot(
       syncEngineImpl,
       query,
       targetId,
       status === 'current',
       targetData.resumeToken
     );
+
```
Comment (milaGGL) [inline packages/firestore/src/core/sync_engine_impl.ts:332]: Do local processing first to get expectedCount, then send Target with listen request.

File: packages/firestore/test/unit/specs/spec_builder.ts [hunk 2]
```diff
@@ -1095,9 +1097,13 @@ export class SpecBuilder {
   private addQueryToActiveTargets(
```
Comment (milaGGL) [inline packages/firestore/test/unit/specs/spec_builder.ts:1097]: add nullable expectedCount parameter to current spec tests.

File: packages/firestore/test/unit/specs/listen_spec.test.ts [hunk 3]
```diff
@@ -1809,4 +1809,40 @@ describeSpec('Listens:', [], () => {
...
+    const docA = doc('collection/a', 1000, { key: 'a' });
+    const docB = doc('collection/b', 1000, { key: 'b' });
+    return spec()
+      .withGCEnabled(false)
+      .userListens(query1)
+      .watchAcksFull(query1, 1000, docA, docB)
+      .expectEvents(query1, { added: [docA, docB] })
+      .userUnlistens(query1)
+      .userDeletes('collection/a')
+      .userListens(query1, { resumeToken: 'resume-token-1000' }, 2)
```
Comment (milaGGL) [inline packages/firestore/test/unit/specs/listen_spec.test.ts:1845]: @dconeybe , sth is off here, the getRemoteKeysForTarget is not considering the mutations happened outside the listen. 🤔

File: packages/firestore/test/unit/specs/spec_builder.ts [hunk 4]
```diff
@@ -1095,9 +1097,13 @@ export class SpecBuilder {
...
     targetId: number,
     query: Query,
-    resumeToken?: string,
-    readTime?: TestSnapshotVersion
+    resume?: { resumeToken?: string; readTime?: TestSnapshotVersion },
+    expectedCount?: number
   ): void {
```
Comment (milaGGL) [inline packages/firestore/test/unit/specs/spec_builder.ts:1102]: add nullable expectedCount to follow resumeToken/readTime.

File: packages/firestore/src/core/sync_engine_impl.ts [hunk 5]
```diff
@@ -331,6 +328,10 @@ export async function syncEngineListen(
...
+    if (syncEngineImpl.isPrimaryClient) {
```
Comment (dconeybe) [inline packages/firestore/src/core/sync_engine_impl.ts:332]: Is it actually necessary to move this "chunk" of code around? I had originally thought that the return value of `addLocalQueryTarget()` above could have been used somehow to get the `expectedCount`, avoiding having to call `getRemoteKeysForTarget()` in `remoteStoreListen()`. But that appears to not be the case, in which case it makes sense to still send off query to the backend before computing the local snapshot.

File: packages/firestore/src/local/target_data.ts [hunk 6]
```diff
@@ -78,17 +83,19 @@ export class TargetData {
...
-   * Creates a new target data instance with an updated resume token and
-   * snapshot version.
+   * Creates a new target data instance with an updated resume token,
+   * snapshot version, and expectedCount if presented.
    */
   withResumeToken(
     resumeToken: ByteString,
-    snapshotVersion: SnapshotVersion
+    snapshotVersion: SnapshotVersion,
+    expectedCount?: number
```
Comment (dconeybe) [inline packages/firestore/src/local/target_data.ts:98]: I don't really like having `expectedCount` be optional here. It makes it too easy to forget to specify it. Is it difficult to drop the `?` and make it a required argument?

File: packages/firestore/test/unit/specs/spec_builder.ts [hunk 7]
```diff
@@ -529,19 +535,19 @@ export class SpecBuilder {
   expectActiveTargets(
     ...targets: Array<{
       query: Query;
-      resumeToken?: string;
-      readTime?: TestSnapshotVersion;
+      resume?: { resumeToken?: string; readTime?: TestSnapshotVersion };
```
Comment (dconeybe) [inline packages/firestore/test/unit/specs/spec_builder.ts:538]: Why did you combine the `resumeToken` and `readTime` into their own, dedicated object? I may be missing something, but I don't see the purpose in this and it adds a lot of seemingly-unnecessary changes to this PR's diff.

File: packages/firestore/test/unit/specs/spec_builder.ts [hunk 8]
```diff
@@ -1095,9 +1097,13 @@ export class SpecBuilder {
   private addQueryToActiveTargets(
     targetId: number,
     query: Query,
-    resumeToken?: string,
-    readTime?: TestSnapshotVersion
+    resume?: { resumeToken?: string; readTime?: TestSnapshotVersion },
+    expectedCount?: number
   ): void {
+    if (!(resume?.resumeToken || resume?.readTime) && expectedCount) {
+      fail('ExpectedCount is present wihtout a resumeToken or readTime.');
```
Comment (dconeybe) [inline packages/firestore/test/unit/specs/spec_builder.ts:1104]: nit: spelling error: "wihtout"

File: packages/firestore/test/unit/specs/spec_builder.ts [hunk 9]
```diff
@@ -1095,9 +1097,13 @@ export class SpecBuilder {
   private addQueryToActiveTargets(
     targetId: number,
     query: Query,
-    resumeToken?: string,
-    readTime?: TestSnapshotVersion
+    resume?: { resumeToken?: string; readTime?: TestSnapshotVersion },
+    expectedCount?: number
```
Comment (dconeybe) [inline packages/firestore/test/unit/specs/spec_builder.ts:1101]: Consider adding `expectedCount` into the `resume` object, instead of having it as a separate argument.

File: packages/firestore/test/unit/specs/spec_builder.ts [hunk 10]
```diff
@@ -1095,9 +1097,13 @@ export class SpecBuilder {
   private addQueryToActiveTargets(
     targetId: number,
     query: Query,
-    resumeToken?: string,
-    readTime?: TestSnapshotVersion
+    resume?: { resumeToken?: string; readTime?: TestSnapshotVersion },
```
Comment (dconeybe) [inline packages/firestore/test/unit/specs/spec_builder.ts:1100]: Since this "resume" type is being used in at least two places in this file, consider adding an `interface` declaration for it.

**Review by dconeybe — CHANGES_REQUESTED** [review]
Thanks for working on this! I'm still diving into some of the details, but at a high level it looks great. Here is my first round of feedback.

File: packages/firestore/src/core/sync_engine_impl.ts (hunk 5 above)
Comment (dconeybe) [reply packages/firestore/src/core/sync_engine_impl.ts:332]: After discussion, it is indeed required to call `remoteStoreListen()` _after_ `initializeViewAndComputeSnapshot()` because the latter initializes some data in the sync engine that `getRemoteKeysForTarget()` uses, namely `queriesByTarget` and `queryViewsByQuery`.

File: packages/firestore/test/unit/specs/listen_spec.test.ts [hunk 11]
```diff
@@ -1809,4 +1809,94 @@ describeSpec('Listens:', [], () => {
...
+            resumeToken: 'resume-token-2000',
+            expectedCount: 2
+          })
+          .expectEvents(query1, { added: [docA, docB], fromCache: true })
+      );
+    }
+  );
+
+  specTest(
+    'ExpectedCount should equal to the number of documents that last matched the query at the resume token',
```
Comment (dconeybe) [inline packages/firestore/test/unit/specs/listen_spec.test.ts:1848]: How about this test name:

Resuming a query should specify expectedCount that does not include pending mutations

File: packages/firestore/test/unit/specs/listen_spec.test.ts [hunk 12]
```diff
@@ -1809,4 +1809,94 @@ describeSpec('Listens:', [], () => {
...
+    }
+  );
+
+  specTest(
+    'ExpectedCount in listen request should work after coming back online',
+    [],
+    () => {
+      const query1 = query('collection');
+      const docA = doc('collection/a', 1000, { key: 'a' });
+      const docBLocal = doc('collection/b', 1000, {
```
Comment (dconeybe) [inline packages/firestore/test/unit/specs/listen_spec.test.ts:1882]: Please remove the local mutations from this test. The local mutations scenario is verified in the previous test and it just adds noise to this test.

**Review by dconeybe — CHANGES_REQUESTED** [review]
Just a few minor comments, then LGTM!

File: packages/firestore/src/local/target_data.ts [hunk 13]
```diff
@@ -97,7 +103,24 @@ export class TargetData {
       this.sequenceNumber,
       snapshotVersion,
       this.lastLimboFreeSnapshotVersion,
-      resumeToken
+      resumeToken,
```
[the remark this answers is not in the document]
Comment (milaGGL) [reply packages/firestore/src/local/target_data.ts:106]: Good idea. If we are doing this, then we need to update the spec test builder/runner, as the activeTargets in spec is added on listen request, and wouldn't be modified when we received new resume tokens.

[the remark this answers is not in the document]
Comment (milaGGL) [reply packages/firestore/src/local/target_data.ts:106]: Screenshot 2022-12-21 at 12 45 55 PM

[the remark this answers is not in the document]
Comment (milaGGL) [reply packages/firestore/src/local/target_data.ts:106]: So the refreshed resumeToken/expectedCount in TargetData while listening is not related to that in Target we created on listen. 
So the **validateActiveTargets() in each step** totally throws me off, as it is always checking the Target that was added onListen step and never modified, or removed that unlisten step.

[the remark this answers is not in the document]
Comment (milaGGL) [reply packages/firestore/src/local/target_data.ts:106]: Other than spec test, resetting the TargetData.expectedCount to default 0 should not make a big impact on anything else, as we are not syncing expectedCount on watch change, and when creating a Target, we always recalculate the expectedCount.

File: packages/firestore/src/local/target_data.ts [hunk 14]
```diff
@@ -104,7 +104,7 @@ export class TargetData {
       snapshotVersion,
       this.lastLimboFreeSnapshotVersion,
       resumeToken,
-      this.expectedCount
+      /** expectedCount= */ null
```
Comment (dconeybe) [inline packages/firestore/src/local/target_data.ts:107]: nit: `/**` -> `/*`. The double-asterisk is used for method/class documentation. For a normal comment just use one asterisk.

File: packages/firestore/src/local/target_data.ts [hunk 15]
```diff
@@ -66,7 +66,12 @@ export class TargetData {
      * matches the target. The resume token essentially identifies a point in
      * time from which the server should resume sending results.
      */
-    readonly resumeToken: ByteString = ByteString.EMPTY_BYTE_STRING
+    readonly resumeToken: ByteString = ByteString.EMPTY_BYTE_STRING,
+    /**
+     * The number of documents that last matched the query at the resume token or
+     * read time.
```
Comment (dconeybe) [inline packages/firestore/src/local/target_data.ts:72]: Add a sentence to explain what `expectedCount==null` means.

File: packages/firestore/test/unit/specs/listen_spec.test.ts [hunk 16]
```diff
@@ -1809,4 +1809,86 @@ describeSpec('Listens:', [], () => {
...
+
+      return (
+        spec()
+          .withGCEnabled(false)
+          .userListens(query1)
+          .watchAcksFull(query1, 1000)
+          .expectEvents(query1, {})
+          .userUnlistens(query1)
+          .watchRemoves(query1)
+          // There is 0 remote document from previous listen.
```
Comment (dconeybe) [inline packages/firestore/test/unit/specs/listen_spec.test.ts:1829]: nit: "There is" -> "There are"

File: packages/firestore/test/unit/specs/listen_spec.test.ts [hunk 17]
```diff
@@ -1809,4 +1809,86 @@ describeSpec('Listens:', [], () => {
...
+
+      return spec()
+        .withGCEnabled(false)
+        .userListens(query1)
+        .watchAcksFull(query1, 1000, docA)
+        .expectEvents(query1, { added: [docA] })
+        .disableNetwork()
+        .expectEvents(query1, { fromCache: true })
+        .enableNetwork()
+        .restoreListen(query1, 'resume-token-1000', 1);
```
Comment (dconeybe) [inline packages/firestore/test/unit/specs/listen_spec.test.ts:1891]: Is the `1` argument the expected count? If so, please add a `/*expectedCount=*/` comment here to clarify.

File: packages/firestore/test/unit/specs/spec_builder.ts [hunk 18]
```diff
@@ -1095,9 +1093,12 @@ export class SpecBuilder {
   private addQueryToActiveTargets(
     targetId: number,
     query: Query,
-    resumeToken?: string,
-    readTime?: TestSnapshotVersion
+    resume?: ResumeSpec
   ): void {
+    if (!(resume?.resumeToken || resume?.readTime) && resume?.expectedCount) {
+      fail('ExpectedCount is present without a resumeToken or readTime.');
```
Comment (dconeybe) [inline packages/firestore/test/unit/specs/spec_builder.ts:1099]: nit: "ExpectedCount" -> "expectedCount" so that the character cases are consistent.

**Review by dconeybe — CHANGES_REQUESTED** [review]
Just a few more very minor comments.

**Review by dconeybe — APPROVED** [review]
